solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -491,6 +494,41 @@ public boolean isClosed() {
     assert ObjectReleaseTracker.track(this);
   }
 
+  /**
+   * <p>Verifies if /clusterstate.json exists in Zookeepeer, and if it does and is not empty, refuses to start and outputs
+   * a helpful message regarding collection migration.</p>
+   *
+   * <p>If /clusterstate.json exists and is empty, it is removed.</p>
+   */
+  private void checkNoOldClusterstate(final SolrZkClient zkClient) throws InterruptedException {
+    try {
+      if (!zkClient.exists(ZkStateReader.UNSUPPORTED_CLUSTER_STATE, true)) {
+        return;
+      }
+
+      final byte[] data = zkClient.getData(ZkStateReader.UNSUPPORTED_CLUSTER_STATE, null, null, true);
+
+      if (Arrays.equals("{}".getBytes(StandardCharsets.UTF_8), data)) {
+        // Empty json. This log will only occur once.
+        log.warn("{} no longer supported starting with Solr 9. Found empty file on Zookeeper, deleting it.", ZkStateReader.UNSUPPORTED_CLUSTER_STATE);
+        zkClient.delete(ZkStateReader.UNSUPPORTED_CLUSTER_STATE, -1, true);
```

Comment:
I don't think that's an issue: if /clusterstate.json is non empty, no node running 9.0 will start.
If /clusterstate.json exists and is empty, the first starting node on 9.0 will delete it as it starts.
8.x nodes that might start afterwards (who knows) will I believe recreate the file (that will be deleted again when a 9.0 node starts).

Reply:
Running 8.x nodes might get a callback since they have a watch. Have not checked the callback handling code, but if the code don't handle DELETE operation then either nothing happens or an exception is thrown...

Reply:
Looking at pre PR master branch.

The watcher on /clusterstate.json is an instance of LegacyClusterStateWatcher (subclass of ZkStateReader).
The watcher processing is done in refreshAndWatch() that calls ZkStateReader.refreshLegacyClusterState() and does some exception handling.

Even though refreshAndWatch() handles KeeperException.NoNodeException by throwing a SolrException SERVICE_UNAVAILABLE, this never happens: refreshLegacyClusterState() catches that exception, a comment says "Ignore missing legacy clusterstate.json." and the catch builds what would be an empty clusterstate.

We should be fine.